### PR TITLE
Add Scaffold Android Library Project

### PR DESCRIPTION
**Pull request by omuomugin**
## What
This is very first step for adding plasma android sdk.
Right now it is only doing "hello world".

## Task List
- [x] Add Scaffold Project
- [x] Documenting to README so that every one should understand how to start and how to build

README is here
https://github.com/cryptoeconomicslab/plasma-android-sdk/blob/219a9758fa4d449657b79289cd9f7d66a6c8a838/README.md

## TODOS
Create issues for following todos
- [x] modify travis ci file due to changing package structure -> implement in other PR 
- [x] figure it out why rust building fail and fix it
- [x] figure it out why cannot run on Android emulator and fix it
- [x] create package for communication layer
- [x] create package for library layer

### CODE REVIEW

File: rust/src/lib.rs [hunk 1]
```diff
@@ -1,14 +1,15 @@
 #![cfg(target_os = "android")]
 #![allow(non_snake_case)]
 
-use ethereum_types::Address;
+// FIXME: add package below, for now some how failed to build
```
Comment (omuomugin) [inline rust/src/lib.rs]: FIXME : this disables from building so just commenting out for now.

File: rust/src/lib.rs [hunk 2]
```diff
@@ -1,14 +1,15 @@
...
+// use ethereum_types::Address;
 use jni::objects::{JObject, JString};
 use jni::sys::jstring;
 use jni::JNIEnv;
 use std::ffi::{CStr, CString};
 
 #[no_mangle]
-pub unsafe extern "C" fn Java_com_example_android_MainActivity_hello(
+pub unsafe extern "C" fn Java_com_cryptoeconomicslab_plasma_1android_1sdk_ExternalMethodsKt_hello(
```
Comment (omuomugin) [inline rust/src/lib.rs]: [MEMO]
This is the tough part and we must hard code the signature of method we are using in Kotlin.

File: settings.gradle [hunk 3]
```diff
@@ -0,0 +1 @@
+include ':demo-hello-world', ':plasma_android_sdk', ':demo-payment-usecase'
```
Comment (omuomugin) [inline settings.gradle]: [MEMO]
write now we have library module and 2 application module for sample

File: rust/tools/build.sh [hunk 4]
```diff
@@ -0,0 +1,14 @@
+#!/usr/bin/env bash
```
Comment (omuomugin) [inline rust/tools/build.sh]: [MEMO]
This script is for local and it will build and copy those binary to library project

**Comment by omuomugin** [comment]
Although there are some todos, it is ready for review now

todos are
- fix travis ci (or just remove it for now)
- adding comment on source code (I want to know where it hard to understand first)

**Comment by omuomugin** [comment]
Also it is a known issue but it does not work with emulator (maybe there are setting for using NDK in emulator). 

I saw it working on Pixel2

File: plasma_android_sdk/src/main/java/com/cryptoeconomicslab/plasma_android_sdk/ExternalMethods.kt [hunk 5]
```diff
@@ -0,0 +1,3 @@
+package com.cryptoeconomicslab.plasma_android_sdk
+
+internal external fun hello(string: String): String
```
Comment (omuomugin) [inline plasma_android_sdk/src/main/java/com/cryptoeconomicslab/plasma_android_sdk/ExternalMethods.kt]: This file can be more refactored but for now I'll leave with this.
I'll do that in different PR.

File: rust/src/lib.rs (hunk 1 above)
Comment (tkmct) [reply rust/src/lib.rs]: in some point we gotta figure out why it's failing.

Comment (omuomugin) [reply rust/src/lib.rs]: I'll be doing that next in smaller PR.

File: README.md [hunk 6]
```diff
@@ -1,13 +1,68 @@
...
+└── README.md
+```
+
+## Architecture
+![](assets/architecture.jpeg)
+
+## How to start
+Usually you only need to do this process just once.
+
+### Overview
```
Comment (syuhei176) [inline README.md]: Do we need to add target with rustup?

```
rustup target add aarch64-linux-android armv7-linux-androideabi i686-linux-android
```

Comment (omuomugin) [reply README.md]: My reference for doing this was this blog post

https://mozilla.github.io/firefox-browser-architecture/experiments/2017-09-21-rust-on-android.html

and in here it says 

> Let’s go ahead and add our newly created Android architectures to rustup so we can use them during cross compilation:

But don't actually know we need this or not.

Comment (omuomugin) [reply README.md]: Could you try building android app without `rustup` ?
I was able to build this repository from clean environment with this README instruction.

Comment (syuhei176) [reply README.md]: Really? I couldn't. and got errors below. After I add targets, works fine to me.

```
error[E0463]: can't find crate for `core`
  |
  = note: the `i686-linux-android` target may not be installed

error: aborting due to previous error

For more information about this error, try `rustc --explain E0463`.
error[E0463]: can't find crate for `core`
  |
  = note: the `i686-linux-android` target may not be installed

error: aborting due to previous error

For more information about this error, try `rustc --explain E0463`.
error: Could not compile `void`.
warning: build failed, waiting for other jobs to finish...
error: Could not compile `cfg-if`.
warning: build failed, waiting for other jobs to finish...
error[E0463]: can't find crate for `std`
  |
  = note: the `i686-linux-android` target may not be installed
```

Comment (syuhei176) [reply README.md]: > My reference for doing this was this blog post

> https://mozilla.github.io/firefox-browser-architecture/experiments/2017-09-21-rust-on-android.html

> and in here it says

OK I will check!

Comment (omuomugin) [reply README.md]: Oh wait it is my bad.
We need `rustup target add aarch64-linux-android armv7-linux-androideabi i686-linux-android` and I thought I put that in to README (but not).

Sorry to misunderstand contex, I'll fix it.

Comment (syuhei176) [reply README.md]: no problem!

**Comment by syuhei176** [comment]
It's so great and amazing PR! I can see almost all the contents which we want. You can fix CI removing `deploy` section. Looks very good to me!

**Comment by omuomugin** [comment]
Thanks for code reviews I created issues for the left todos.
Merging my first PR in cryptoeconomicslab 🎉

**Pull request merged**